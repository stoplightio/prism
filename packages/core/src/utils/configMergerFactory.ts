import { PartialPrismConfig, PrismConfig, PrismConfigFactory, resolveConfig } from '..';

const _compact = require('lodash/compact');
const _merge = require('lodash/merge');

/**
 * Merges all passed configs. Each next config wil override each previous config.
 */
export function configMergerFactory<C, I>(
  baseConfig: PrismConfig<C, I>,
  ...configs: Array<PartialPrismConfig<C, I> | undefined>
): PrismConfigFactory<C, I> {
  const factory = async (input: I, defaultConfig?: PartialPrismConfig<C, I>): Promise<C> => {
    const resolvedConfigs =
      // remove any falsy resolved configs
      _compact(
        await Promise.all(
          // remove falsy config props
          _compact([baseConfig, ...configs])
            // resolve each config (resolveConfig is async)
            .map((c: C) => resolveConfig<C, I>(input, c, defaultConfig))
        )
      );

    if (!resolvedConfigs.length) {
      throw new Error('All configurations passed to the factory are undefined.');
    }

    // merge the configs over each other, in order
    return _merge({}, ...resolvedConfigs);
  };
  return factory as PrismConfigFactory<C, I>;
}
