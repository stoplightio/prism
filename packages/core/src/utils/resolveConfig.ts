import { PartialPrismConfig, PartialPrismConfigFactory } from '../types';

export async function resolveConfig<Config, Input>(
  input: Input,
  config: PartialPrismConfig<Config, Input>,
  defaultConfig?: PartialPrismConfig<Config, Input>
): Promise<Partial<Config>> {
  if (typeof config === 'function') {
    // config factory function
    return await (config as PartialPrismConfigFactory<Config, Input>)(input, defaultConfig);
  }
  return config;
}
