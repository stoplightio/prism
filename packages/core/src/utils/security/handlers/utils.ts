import { DiagnosticSeverity } from '@stoplight/types';
import { Either, left, right } from 'fp-ts/lib/Either';
import { IPrismDiagnostic } from '../../../types';

const forbiddenErr: IPrismDiagnostic = {
  code: 403,
  message: 'Invalid credentials used',
  severity: DiagnosticSeverity.Error,
};

const invalidCredsErr = left(forbiddenErr);

export function genRespForScheme<R>(
  isSchemeProper: boolean,
  isCredsGiven: boolean,
  resource: R,
  msg: string,
): Either<IPrismDiagnostic, R> {
  const handler = [
    {
      test: () => isSchemeProper && isCredsGiven,
      handle: () => right(resource),
    },
    {
      test: () => isSchemeProper,
      handle: () => invalidCredsErr,
    },
  ].find(possibleHandler => possibleHandler.test());

  return handler ? handler.handle() : left(genUnauthorisedErr(msg));
}

export const genUnauthorisedErr = (msg: string): IPrismDiagnostic => ({
  severity: DiagnosticSeverity.Error,
  message: 'Invalid security scheme used',
  code: 401,
  tags: msg ? [msg] : [],
});

export function isScheme(shouldBeScheme: string, authScheme: string) {
  return (authScheme || '').toLowerCase() === shouldBeScheme;
}

export function when(condition: boolean, errorMessage: string, resource?: unknown): Either<IPrismDiagnostic, unknown> {
  return condition ? right(resource) : left(genUnauthorisedErr(errorMessage));
}
