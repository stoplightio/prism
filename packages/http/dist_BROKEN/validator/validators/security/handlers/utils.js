"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.when = exports.isScheme = exports.genUnauthorisedErr = exports.genRespForScheme = void 0;
const types_1 = require("@stoplight/types");
const E = require("fp-ts/Either");
const function_1 = require("fp-ts/function");
const function_2 = require("fp-ts/function");
const genRespForScheme = (isSchemeProper, isCredsGiven, msg) => function_1.pipe(isSchemeProper, E.fromPredicate(function_2.identity, () => exports.genUnauthorisedErr(msg)), E.chain(() => exports.when(isCredsGiven, undefined)));
exports.genRespForScheme = genRespForScheme;
const genUnauthorisedErr = (msg) => ({
    severity: types_1.DiagnosticSeverity.Error,
    message: 'Invalid security scheme used',
    code: 401,
    tags: msg ? [msg] : [],
});
exports.genUnauthorisedErr = genUnauthorisedErr;
function isScheme(shouldBeScheme, authScheme) {
    return authScheme.toLowerCase() === shouldBeScheme;
}
exports.isScheme = isScheme;
const when = (condition, errorMessage) => function_1.pipe(condition, E.fromPredicate(function_2.identity, () => exports.genUnauthorisedErr(errorMessage)));
exports.when = when;
