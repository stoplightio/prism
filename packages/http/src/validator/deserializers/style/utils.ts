import { Dictionary } from '@stoplight/types';

export function createObjectFromKeyValList(items: string[]) {
  return items.reduce((obj: Dictionary<unknown>, item, i) => {
    if (i % 2 === 0) {
      obj[item] = undefined;
    } else {
      obj[items[i - 1]] = item;
    }

    return obj;
  }, {});
}
