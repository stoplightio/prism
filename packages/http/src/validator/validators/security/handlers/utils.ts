import { DiagnosticSeverity } from '@stoplight/types';
import { Either, left, right } from 'fp-ts/lib/Either';
import { IPrismDiagnostic } from '@stoplight/prism-core';
import { IHttpRequest } from '../../../../types';

export type ValidateSecurityFn = (
  input: Pick<IHttpRequest, 'headers' | 'url'>,
  name: string
) => Either<IPrismDiagnostic, unknown>;

export function genRespForScheme(
  isSchemeProper: boolean,
  isCredsGiven: boolean,
  msg: string
): Either<IPrismDiagnostic, unknown> {
  if (isSchemeProper) {
    return when(isCredsGiven, undefined);
  }

  return left(genUnauthorisedErr(msg));
}

export const genUnauthorisedErr = (msg?: string): IPrismDiagnostic => ({
  severity: DiagnosticSeverity.Error,
  message: 'Invalid security scheme used',
  code: 401,
  tags: msg ? [msg] : [],
});

export function isScheme(shouldBeScheme: string, authScheme: string) {
  return authScheme.toLowerCase() === shouldBeScheme;
}

export function when(condition: boolean, errorMessage: string | undefined): Either<IPrismDiagnostic, unknown> {
  return condition ? right(true) : left(genUnauthorisedErr(errorMessage));
}
