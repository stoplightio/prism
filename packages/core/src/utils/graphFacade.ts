import { IGraphite } from '@stoplight/graphite';
import {
  createFileSystemBackend,
  FileSystemBackend,
  FilesystemNodeType,
} from '@stoplight/graphite/backends/filesystem';
import { filenameToLanguage } from '@stoplight/graphite/backends/filesystem/utils';
import { ISourceNode, NodeCategory } from '@stoplight/graphite/graph/nodes';
import { createGraphite } from '@stoplight/graphite/graphite';
import { createOas2HttpPlugin } from '@stoplight/graphite/plugins/http/oas2';
import { createOas3HttpPlugin } from '@stoplight/graphite/plugins/http/oas3';
import { createJsonPlugin } from '@stoplight/graphite/plugins/json';
import { createOas2Plugin } from '@stoplight/graphite/plugins/oas2';
import { createOas3Plugin } from '@stoplight/graphite/plugins/oas3';
import { createYamlPlugin } from '@stoplight/graphite/plugins/yaml';
import { IHttpOperation } from '@stoplight/types';
import * as fs from 'fs';
import { resolve } from 'path';

import { compact } from 'lodash';

export class GraphFacade {
  private fsBackend: FileSystemBackend;
  private graphite: IGraphite;

  constructor() {
    const graphite = (this.graphite = createGraphite());
    graphite.registerPlugins(
      createJsonPlugin(),
      createYamlPlugin(),
      createOas2Plugin(),
      createOas3Plugin(),
      createOas2HttpPlugin(),
      createOas3HttpPlugin(),
    );
    this.fsBackend = createFileSystemBackend(graphite, fs);
  }

  public async createFilesystemNode(fsPath: string) {
    const resourceFile = resolve(fsPath);
    const stat = fs.lstatSync(resourceFile);

    if (stat.isDirectory()) {
      this.graphite.graph.addNode({
        category: NodeCategory.Source,
        type: FilesystemNodeType.Directory,
        path: resourceFile,
      });
      this.fsBackend.readdir(fsPath);
    } else if (stat.isFile()) {
      this.graphite.graph.addNode({
        category: NodeCategory.Source,
        type: FilesystemNodeType.File,
        language: filenameToLanguage(resourceFile),
        path: resourceFile,
      });
      this.fsBackend.readFile(resourceFile);
    }

    await this.graphite.scheduler.drain();
  }

  public async createRawNode(raw: string, { type, path }: Pick<ISourceNode, 'type' | 'path'>) {
    this.graphite.graph.addNode({
      category: NodeCategory.Source,
      type,
      language: filenameToLanguage(path),
      path: '/' + path,
      data: { raw },
    });

    await this.graphite.scheduler.drain();
  }

  get httpOperations(): IHttpOperation[] {
    const nodes = this.graphite.graph.virtualNodes.filter(node => node.type === 'http_operation');
    return compact(nodes.map<IHttpOperation>(node => node.data as IHttpOperation));
  }
}
