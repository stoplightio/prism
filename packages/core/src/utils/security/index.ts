import { DiagnosticSeverity } from '@stoplight/types';
import { Either, fold, isLeft, isRight, Left, left } from 'fp-ts/lib/Either';
import { none, Option, some } from 'fp-ts/lib/Option';
import { pipe } from 'fp-ts/lib/pipeable';
import { flatten, get, identity } from 'lodash';
import { noop, set } from 'lodash/fp';
import { IPrismDiagnostic } from '../../types';
import { securitySchemeHandlers } from './handlers';
import { SecurityScheme } from './handlers/types';

function gatherInvalidResults(
  error: Left<IPrismDiagnostic>,
  invalidSecuritySchemes: Array<Array<Either<IPrismDiagnostic, unknown>>>,
) {
  const invalidSecurity = gatherWWWAuthHeader(invalidSecuritySchemes, ['tags'], error.left);
  return some(invalidSecurity);
}

function gatherValidationResults(securitySchemes: SecurityScheme[][], someInput: unknown, resource: unknown) {
  const authResults = getAuthResults(securitySchemes, someInput, resource);

  const validSecurityScheme = authResults.find(authRes => authRes.every(isRight));
  const invalidSecuritySchemes = authResults.filter(authRes => authRes.some(isLeft));

  const firstLeft = invalidSecuritySchemes[0] && invalidSecuritySchemes[0].find(isLeft);

  if (!validSecurityScheme && firstLeft) {
    return gatherInvalidResults(firstLeft, invalidSecuritySchemes);
  } else {
    return none;
  }
}

function gatherWWWAuthHeader(
  authResults: Array<Array<Either<IPrismDiagnostic, unknown>>>,
  pathToHeader: string[],
  firstAuthErr: IPrismDiagnostic,
) {
  const flattenedAuthResults = flatten(authResults);

  if (flattenedAuthResults.length === 1) {
    return firstAuthErr;
  } else {
    const wwwAuthenticateHeaders = flattenedAuthResults.map(authResult =>
      pipe(
        authResult,
        fold(result => result.tags || [], noop),
      ),
    );

    const firstAuthErrWithAuthHeader = set(pathToHeader, flatten(wwwAuthenticateHeaders), firstAuthErr);

    return wwwAuthenticateHeaders.every(identity) ? firstAuthErrWithAuthHeader : firstAuthErr;
  }
}

function getAuthResult(
  firstAuthErrAsLeft: Left<IPrismDiagnostic>,
  authResult: Array<Either<IPrismDiagnostic, unknown>>,
) {
  const firstAuthErr: IPrismDiagnostic = pipe(
    firstAuthErrAsLeft,
    fold<IPrismDiagnostic, IPrismDiagnostic, IPrismDiagnostic>(identity, identity),
  );

  const invalidResultWithAuthHeader = gatherWWWAuthHeader([authResult], ['tags'], firstAuthErr);

  return [left(invalidResultWithAuthHeader)];
}

function getAuthResults(securitySchemes: SecurityScheme[][], someInput: unknown, resource: unknown) {
  return securitySchemes.map(securitySchemePairs => {
    const authResult = securitySchemePairs.map(securityScheme => {
      const schemeHandler = securitySchemeHandlers.find(handler => handler.test(securityScheme));

      return schemeHandler
        ? schemeHandler.handle(someInput, securityScheme.name, resource)
        : left({
            message: 'We currently do not support this type of security scheme.',
            severity: DiagnosticSeverity.Warning,
          });
    });

    const firstAuthErrAsLeft = authResult.find(isLeft);

    if (firstAuthErrAsLeft) {
      return getAuthResult(firstAuthErrAsLeft, authResult);
    } else {
      return authResult;
    }
  });
}

export function validateSecurity(someInput: unknown, resource?: unknown): Option<IPrismDiagnostic> {
  const securitySchemes = get(resource, 'security', []);

  if (!securitySchemes.length) {
    return none;
  } else {
    return gatherValidationResults(securitySchemes, someInput, resource);
  }
}
