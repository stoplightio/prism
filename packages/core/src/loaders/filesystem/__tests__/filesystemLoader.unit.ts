import * as fs from 'fs';
import { FilesystemLoader } from '../';
import { GraphFacade } from '../../../utils/graphFacade';
jest.mock('../../../utils/graphFacade');
jest.mock('fs');

describe('filesystemLoader', () => {
  const fakeHttpOperations = ['a', 'b', 'c'];
  let graphFacadeMock: any;
  let createFileSystemNodeMock: any;

  beforeEach(() => {
    graphFacadeMock = new GraphFacade();
    createFileSystemNodeMock = graphFacadeMock.createFilesystemNode as jest.Mock;
    createFileSystemNodeMock.mockResolvedValue(true);

    Object.defineProperty(graphFacadeMock, 'httpOperations', {
      get: jest.fn().mockReturnValue(fakeHttpOperations),
    });
  });

  test('given no opts should delegate to graph facade for http operations', async () => {
    jest.spyOn(fs, 'existsSync').mockReturnValueOnce(true);
    jest.spyOn(fs, 'statSync').mockReturnValueOnce({ isDirectory: () => false });
    const filesystemLoader = new FilesystemLoader(graphFacadeMock);

    const operations = await filesystemLoader.load();

    expect(createFileSystemNodeMock).toHaveBeenCalledWith('.');
    expect(operations).toBe(fakeHttpOperations);
  });

  test('given opts should delegate to graph facade for http operations', async () => {
    jest.spyOn(fs, 'existsSync').mockReturnValueOnce(true);
    jest.spyOn(fs, 'statSync').mockReturnValueOnce({ isDirectory: () => false });

    const filesystemLoader = new FilesystemLoader(graphFacadeMock);

    const operations = await filesystemLoader.load({ path: 'a path' });

    expect(createFileSystemNodeMock).toHaveBeenCalledWith('a path');
    expect(operations).toBe(fakeHttpOperations);
  });

  it('throws error when supplied with non-existing path', async () => {
    jest.spyOn(fs, 'existsSync').mockReturnValueOnce(true);
    jest.spyOn(fs, 'statSync').mockReturnValueOnce({ isDirectory: () => true });
    const filesystemLoader = new FilesystemLoader(graphFacadeMock);
    return expect(filesystemLoader.load({ path: 'a path' })).rejects.toThrowErrorMatchingSnapshot();
  });

  it('throws error when supplied with a directory', async () => {
    jest.spyOn(fs, 'existsSync').mockReturnValueOnce(false);
    const filesystemLoader = new FilesystemLoader(graphFacadeMock);
    return expect(filesystemLoader.load({ path: 'a path' })).rejects.toThrowErrorMatchingSnapshot();
  });
});
