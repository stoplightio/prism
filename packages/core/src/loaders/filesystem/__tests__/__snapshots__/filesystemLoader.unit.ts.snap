// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filesystemLoader throws error when supplied with a directory 1`] = `"Non-existing path to spec supplied: a path"`;

exports[`filesystemLoader throws error when supplied with non-existing path 1`] = `"Supplied spec path points to directory. Only files are supported."`;
